Add example of chaining Results using verbose or backpassing syntax

ageron opened the pull request:
The example parses strings such as "John Lennon was born in 1940" to extract the first name, last name, and birth year. It does so first in a verbose manner, using nested `with` blocks, and then uses backpassing instead, which of course is much shorter & cleaner. I also included two commented out functions showing what the upcoming `?` operator will look like.

As I'm quite new to Roc, please let me know if any of this is wrong or not idiomatic, I'm happy to make changes if needed.

bhansconnect commented on examples/Results/main.roc:
```diff
@@ -0,0 +1,106 @@
+# This function parses a string formatted like "{FirstName} {LastName} was born in {Year}"
+# This is the most verbose version. It's a bit too nested, we will do better below.
+parsePersonVerbose = \line ->
+    when line |> Str.split " was born in " is
```
Can we change all the use of `Str.split` to `Str.splitFirst`? Want to make sure we recommend the best api here. It still return s a result, so should be good for showing err handling

bhansconnect replied:
Also should mean we can remove `toPair` below, which I think will help cleanup the example.

It will just be something like:
{before: firstName, after: bithYearStr} <- Str.splitFirst line " was born in " |> Result.try

ageron replied:
Great recommendations, thanks. This is fixed now.

bhansconnect commented on examples/Results/main.roc:
```diff
@@ -0,0 +1,106 @@
+    (fullName, birthYearStr) <- line
+        |> Str.split " was born in "
+        |> toPair
+        |> Result.mapErr (\_ -> Err (ParsingErr InvalidRecordFormat))
```
I would remove `ParsingErr` from all of these errors, I don't think it is idiomatic.

If a user is directly handling the error, it is just noise.
If a user is not directly handling they error, they can easily decide to wrap it:
`parePersonBackpassingV2 line |> Result.mapErr ParsingErr`

Oh, also, you are adding a redundant `Err`. I think this would result in a `Err (Err (ParsingErr ...))` on failure.

ageron replied:
Good points, thanks! Fixed.

bhansconnect commented on examples/Results/main.roc:
@rtfeldman what is your thought on adding backpassing examples? It is deprecated, so maybe we should avoid suggesting it's use?

rtfeldman replied:
Yeah I agree, let's not introduce new uses of backpassing in examples. 👍

bhansconnect replied:
Cool, so then for this example, I think it would be good to have:
1. The raw `when .. is` version
2. A regular lambda version with nesting with/without error handling (equivalent to the backpassing examples but without the backpasssing)
3. Coming soon, `?` exmaples that are commented out.

bhansconnect commented on examples/Results/main.roc:
```diff
@@ -0,0 +1,106 @@
+
+# This function parses a string formatted like "{FirstName} {LastName} was born in {Year}"
+# This is the most verbose version. It's a bit too nested, we will do better below.
+parsePersonVerbose = \line ->
```
Can we add a version of this use `Result.try` and `Result.mapErr`, but no backpassing? I think that will be useful to new users.

ageron replied:
Good suggestion, thanks. I've added `parseWithTryV2`.

ageron commented:
Thanks for all the great feedback!
I'll do this today.

ageron commented:
Done! As you suggested:
* Replaced `Str.split` with `Str.splitFirst`
* Removed `toPair` which is not needed anymore
* Removed the backpassing examples and replaced them with regular `Result.try` using lambdas
* Cleaned up the code and comments, in particular the `?` operator examples.

I can't wait for the `?` operator, it will make Roc code even nicer! 👍

bhansconnect reviewed (approved):
Yeah, this looks great!

bhansconnect commented on ci_scripts/expect_scripts/Results.exp:
```diff
@@ -0,0 +1,17 @@
+
+set timeout 7
+
+spawn ./examples/HelloWorld/main
```
```suggestion
spawn ./examples/Results/main
```

ageron commented:
I had forgotten to run `roc format main.roc`, so I just did that.
Oddly, the tests fail on `spawn ./examples/Arithmetic/main 20 4`, but this PR doesn't touch that so I'm confused.

Anton-4 commented:
> Oddly, the tests fail on spawn ./examples/Arithmetic/main 20 4, 

I'll check that out

Anton-4 commented:
The failure is a segmentation fault.

I wanted to check it out with valgrind but it does not recognize an instruction:
```
❯ valgrind ./examples/Arithmetic/main 20 4
==9642== Memcheck, a memory error detector
==9642== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==9642== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==9642== Command: ./examples/Arithmetic/main 20 4
==9642== 
vex amd64->IR: unhandled instruction bytes: 0xF4 0x8A 0x4F 0xFD 0xF 0xB7 0x57 0xFE 0x41 0x89
vex amd64->IR:   REX=0 REX.W=0 REX.R=0 REX.X=0 REX.B=0
vex amd64->IR:   VEX=0 VEX.L=0 VEX.nVVVV=0x0 ESC=NONE
vex amd64->IR:   PFX.66=0 PFX.F2=0 PFX.F3=0
==9642== valgrind: Unrecognised instruction at address 0x2a7fd3.
==9642==    at 0x2A7FD3: ??? (in /home/username/gitrepos/forks/ageron/roc-examples/examples/Arithmetic/main)
==9642==    by 0x2A83B0: ??? (in /home/username/gitrepos/forks/ageron/roc-examples/examples/Arithmetic/main)
==9642==    by 0x3: ???
==9642==    by 0x480425F: ???
==9642==    by 0x7FFFFFFFFFFFFFFF: ???
==9642==    by 0x480461F: ???
==9642==    by 0x2: ???
==9642==    by 0x2AC232: ??? (in /home/username/gitrepos/forks/ageron/roc-examples/examples/Arithmetic/main)
==9642==    by 0x2A7845: ??? (in /home/username/gitrepos/forks/ageron/roc-examples/examples/Arithmetic/main)
==9642==    by 0x1FFEFFF8DF: ???
==9642==    by 0x2: ???
==9642==    by 0x34BC73: ??? (in /home/username/gitrepos/forks/ageron/roc-examples/examples/Arithmetic/main)
==9642== Your program just tried to execute an instruction that Valgrind
==9642== did not recognise.  There are two possible reasons for this.
==9642== 1. Your program has a bug and erroneously jumped to a non-code
==9642==    location.  If you are running Memcheck and you just saw a
==9642==    warning about a bad jump, it's probably your program's fault.
==9642== 2. The instruction is legitimate but Valgrind doesn't handle it,
==9642==    i.e. it's Valgrind's fault.  If you think this is the case or
==9642==    you are not sure, please let us know and we'll try to fix it.
==9642== Either way, Valgrind will now raise a SIGILL signal which will
==9642== probably kill your program.
==9642== 
==9642== Process terminating with default action of signal 4 (SIGILL)
==9642==  Illegal opcode at address 0x2A7FD3
```
I tried several ways to figure out the actual instruction but failed, can you give me tip here @bhansconnect?

bhansconnect commented:
I would open it up in objdump and jump to those instructions. See what it says. Is Val grind somehow in the middle of an instruction. Are those bytes garbage. Other.

Anton-4 commented:
The arithmetic example runs fine with basic-cli 0.13 so I'm just going to focus on #192.
I've attached my attempts to view the unrecognized instruction just in case we need it in the future:
[unrecognised_instruction.txt](https://github.com/user-attachments/files/16561109/unrecognised_instruction.txt)

Anton-4 commented:
https://github.com/roc-lang/examples/pull/203 needs to be merged in first to resolve the expected checks.

Anton-4 commented:
> A new operator `?` (the "try" operator) will soon be added to the language

? is here now, we should update this before merging

ageron commented:
I'll update this PR today 👍

Anton-4 reviewed (approved):
Thanks @ageron!
